src/extensions/rppasio/rppasio/schedulers/strand.hpp

```diff
@@ -0,0 +1,195 @@
+                if (handler.is_disposed())
+                    return;
+
+                auto timer = std::make_shared<asio::basic_waitable_timer<rpp::schedulers::clock_type>>(m_strand.context(), tp);
+                timer->async_wait(asio::bind_executor(m_strand, [self = this->shared_from_this(), timer, fn = std::forward<Fn>(fn), handler = std::forward<Handler>(handler), ... args = std::forward<Args>(args)](const asio::error_code& ec) mutable {
+                    if (ec || handler.is_disposed())
+                        return;
+
+                    current_thread_queue_guard guard{*self};
+                    if (const auto new_duration = fn(handler, args...))
+                        self->defer_to(new_duration->value, std::move(fn), std::move(handler), std::move(args)...);
+                }));
+            }
+
+        private:
+            // Guard draining schedulables queued to thread local queue to schedule them back to strand queue
+            class current_thread_queue_guard
+            {
+            public:
+                current_thread_queue_guard(const state_t& state)
+                    : m_process_on_destruction{!rpp::schedulers::current_thread::get_queue()}
+                    , m_state(state)
+                {
+                    if (m_process_on_destruction)
+                        rpp::schedulers::current_thread::get_queue() = &m_queue;
+                }
+                ~current_thread_queue_guard()
+                {
+                    if (m_process_on_destruction)
+                        process_queue();
+                }
+                current_thread_queue_guard(const current_thread_queue_guard&) = delete;
+                current_thread_queue_guard(current_thread_queue_guard&&)      = delete;
+
+            private:
+                struct handler
+                {
+                    bool is_disposed() const noexcept
+                    {
+                        return m_schedulable->is_disposed();
+                    }
+
+                    void on_error(const std::exception_ptr& ep) const
+                    {
+                        m_schedulable->on_error(ep);
+                    }
+
+                    std::shared_ptr<rpp::schedulers::details::schedulable_base> m_schedulable;
+                };
+
+            private:
+                void process_queue()
+                {
+                    while (!m_queue.is_empty())
+                    {
+                        const auto top = m_queue.pop();
+                        if (top->is_disposed())
+                            continue;
+
+                        m_state.defer_to(
+                            top->get_timepoint(),
+                            [top](const auto&) -> rpp::schedulers::optional_delay_to {
+                                if (const auto advanced_call = top->make_advanced_call())
+                                {
+                                    top->set_timepoint(worker_strategy::now());
```

Comment:
@victimsnino a bit uncertain if I should set `now()` here

Reply:
you can omit this call at all, handle_advanced_call should handle it by its own

Reply:
but then what do I return as `delay_to`?

Reply:
done